[IRCE] Check loop clone legality before attempting to do so

Keno opened the pull request:
Not all loop bodies are safe to clone. For example, they may have `noduplicate` call in them. Additionally, if the transformation may cause any of the cloned loop bodies to execute conditionally, we must also disallow `convergent` calls, as well as any call that returns a token that is used outside of the loop. The loop unswitching pass had an appropriate legality check for this, but IRCE did not, causing issues downstream (https://github.com/JuliaLang/julia/issues/48918).

As a particular example, consider the test case in this commit, where `opt -passes=irce`, creates invalid IR:
```
opt --passes=irce llvm/test/Transforms/IRCE/pre_post_loops_clone.ll
WARNING: You're attempting to print out a bitcode file.
This is inadvisable as it may cause display problems. If
you REALLY want to taste LLVM bitcode first-hand, you
can force output with the `-f' option.

Instruction does not dominate all uses!
  %token = call token @llvm.source_token()
  call void @llvm.sink_token(token %token)
Instruction does not dominate all uses!
  %token = call token @llvm.source_token()
  call void @llvm.sink_token(token %token)
LLVM ERROR: Broken module found, compilation aborted!
```

Fix this by factoring out the legality check from the loop unswitch pass and using it in IRCE as well. I personally don't have any code that is `noduplicate` or `convergent`, but I don't see any reason why IRCE would be different here.

I will also mention #56243 as related, which contains some related discussion of LCSSA in the presence of tokens.

Fixes #63984

vtjnash commented:
LGTM. Though I would note the description makes it sound NFC for the existing loop unstitching, but it looks like you also improved the used outside BB check to be a more permissive used outside loop check, which should perhaps be noted in the commit message

Keno commented:
@xortator @aleks-tmb Any comments on this? Would be good to get input from people that have touched IRCE in the past. I would like to get this in to address the compiler error. I guess I'll wait another few days, otherwise, I'll make the commit message adjustment that @vtjnash requested and get this in.

Keno commented on llvm/lib/Transforms/Utils/LoopConstrainer.cpp:
```diff
@@ -755,6 +755,13 @@ bool LoopConstrainer::run() {
+      !OriginalLoop.isSafeToCloneConditionally(DT)) {
+    LLVM_DEBUG(dbgs() << "cannot clone loop " << OriginalLoop.getName()
+                      << " because it is not safe to clone.\n");
+    return false;
```
GPT 5.6 points out that this (and the existing bail paths below) are illegal since SCEV has already modified the function here and we're claiming to preserve analysis in this case. Will fix.

Keno commented on llvm/lib/Analysis/LoopInfo.cpp:
```diff
@@ -520,6 +527,28 @@ bool Loop::isSafeToClone() const {
+      return false;
+
+    for (Instruction &I : *BB) {
+      if (I.getType()->isTokenTy()) {
```
GPT 5.6 points out that the pendancy of this PR has crossed the introduction of `isTokenLikeTy`, so this needs updating. Will fix as well.

aleks-tmb commented:
@Keno hi,
Apologies for the delayed response and for being inactive upstream. Let me take a look now

aleks-tmb commented on llvm/lib/Analysis/LoopInfo.cpp:
```diff
@@ -449,6 +449,27 @@ bool Loop::isCanonical(ScalarEvolution &SE) const {
   return true;
 }
 
+static bool loopContainsUser(const Loop &L, const BasicBlock &BB, const Use &U,
```
Could we extract `loopContainsUser` helper in a separate NFC patch to minimize the diff here and make the functional changes easier to review?

Keno replied:
https://github.com/llvm/llvm-project/pull/211087

aleks-tmb commented on llvm/lib/Transforms/Utils/LoopConstrainer.cpp:
```diff
@@ -128,9 +128,10 @@ static const SCEV *getNarrowestLatchMaxTakenCountEstimate(ScalarEvolution &SE,
 
 std::optional<LoopStructure>
-LoopStructure::parseLoopStructure(ScalarEvolution &SE, Loop &L,
+LoopStructure::parseLoopStructure(SCEVExpander &Expander, Loop &L,
```
This change is also a good candidate for extraction into a separate preparatory NFC patch.

Keno replied:
https://github.com/llvm/llvm-project/pull/211089 - not NFC though.

aleks-tmb commented on llvm/lib/Transforms/Scalar/SimpleLoopUnswitch.cpp:
```diff
@@ -3358,19 +3358,10 @@ static bool collectUnswitchCandidatesWithInjections(
-  if (!L.isSafeToClone())
+static bool isSafeForNoNTrivialUnswitching(const DominatorTree &DT, Loop &L,
+                                           LoopInfo &LI) {
+  if (!L.isSafeToCloneConditionally(DT))
```
Would it make sense to split this into two commits?

1. A pure NFC refactor that just moves the existing caller-side checks into the new `isSafeToCloneConditionally` helper, preserving current behavior exactly.
2. A follow-up with the actual semantic changes on top

The current shape reads as NFC at first glance, but there are a couple of subtle behavioral shifts hiding inside - some checks get broader, others get more permissive. Separating them would make the intent easier to follow on review

Keno replied:
Split as requested

aleks-tmb replied:
Thanks! Looks good. 
Could we merge it one by one in separate MRs? For the NFC change, I'm ready to give an immediate approval.

Keno committed:
[IRCE] Check loop clone legality before attempting to do so

Not all loop bodies are safe to clone. For example, they may have
`noduplicate` call in them. Additionally, if the transformation may
cause any of the cloned loop bodies to execute conditionally, we
must also disallow `convergent` calls, as well as any call that
returns a token that is used outside of the loop. The loop unswitching
pass had an appropriate legality check for this, but IRCE did not,
causing issues downstream (https://github.com/JuliaLang/julia/issues/48918).

As a particular example, consider the test case in this commit, where
`opt -passes=irce`, creates invalid IR:
```
opt --passes=irce llvm/test/Transforms/IRCE/pre_post_loops_clone.ll
WARNING: You're attempting to print out a bitcode file.
This is inadvisable as it may cause display problems. If
you REALLY want to taste LLVM bitcode first-hand, you
can force output with the `-f' option.

Instruction does not dominate all uses!
  %token = call token @llvm.source_token()
  call void @llvm.sink_token(token %token)
Instruction does not dominate all uses!
  %token = call token @llvm.source_token()
  call void @llvm.sink_token(token %token)
LLVM ERROR: Broken module found, compilation aborted!
```

Fix this by using the legality check that #217884 and #217916
factored out of the loop unswitch pass (Loop::isSafeToCloneConditionally)
in IRCE as well. I personally don't have any code that is `noduplicate`
or `convergent`, but I don't see any reason why IRCE would be different
here.

I will also mention #56243 as related, which contains some related
discussion of LCSSA in the presence of tokens.

Fixes #63984
